frontend/src/services/authentication_handler_service/authentication_handler_service.ts

```diff
@@ -11,59 +10,134 @@ export class FailureToSignInError extends Error {
+  sub: string,
+}
+
 /**
- * Communicates with gapi to retrieve auth code from user.
- * Exchanges auth code and redirect URI for ID token from backend.
+ * Communicates with gapi to manage a user's login status
+ * and retrieve auth codes / tokens.
  */
 export class AuthenticationHandlerService {
+  token: string | undefined = undefined;
+  authInstance: gapi.auth2.GoogleAuth | undefined = undefined;
+
   /** Backend is responsible for retrieving the ID token. */
-  constructor(private readonly backend: BackendAuthenticationInterface) {
-    // Safe because the gapi script is loaded in the head with no async/defer.
-    // TODO: Make gapi script async and call this when it has loaded
-    window.gapi.load('auth2', () => {
-      window.gapi.auth2.init({
-        client_id: CLIENT_ID,
-      });
-    });
+  constructor(private readonly backend: BackendAuthenticationInterface) {}
+
+  /**
+   * Determines if a user is logged in.
+   * @return true if user is logged in, false otherwise
+   */
+  getUserLoginStatus(): boolean {
+    const auth = this.getAuthInstance();
+    return auth.isSignedIn.get() && this.getToken() !== undefined;
+  }
+
+  /**
+   * Gets the parsed ID token of the user.
+   * @return the parsed token if it exists, otherwise undefined
+   */
+  getParsedToken(): TokenInterface | undefined {
+    const token = this.getToken();
+    return token
+        ? JSON.parse(atob(token.split(".")[1]))
+        : undefined;
+  }
+
+  /**
+   * Gets the ID token of the user.
+   * @return the token if it exists, otherwise undefined
+   */
+  getToken(): string | undefined {
+    const auth = this.getAuthInstance();
+    const response = auth.currentUser.get().getAuthResponse();
+    if (!response.id_token) {
+      return;
+    }
+    return response.id_token;
   }
 
   /**
-   * Signs the user in, retrieving a token from the backend.
+   * Signs the user in, returns the user's ID token.
    * @param scopes the scopes our app requests for oauth
-   * @return the string of the JWT token retrieved by the backend
-   * @throws FailureToSignInError if sign in fails
+   * @return the retrieved ID token
+   * @throws FailureToSignInError if sign in fails (Promise reject)
    */
   async signIn(scopes: string): Promise<string> {
     const code: string | undefined = await this.getAuthCode({scope: scopes});
     if (code === undefined) {
       throw new FailureToSignInError();
     }
-    try {
-      return await this.backend.retrieveToken(code, window.location.origin);
-    } catch (err) {
-      throw new FailureToSignInError();
-    }
+    const redirectUri = window.location.origin;
+    return new Promise(async (resolve, reject) => {
+      try {
+        const token = await this.backend.retrieveToken(code, redirectUri);
+        const signedInToken = this.getToken();
+        this.oneTimeBindToSignIn(token, resolve);
+
+        // If already logged in, listener won't run so we resolve here.
+        if (signedInToken) {
```

Comment:
Move this above the call to oneTimeBindToSignIn, and return. It's confusing if there's a codepath that can potentially call resolve multiple times.

Reply:
That a good point, it has been moved above and a return statement was added. I really appreciate the style feedback here, very helpful to not just know how to do something, but also the right way to.